File: stubs/java/src/main/java/com/beastmode/rmddd/domain/AggregateRoot.java
package com.beastmode.rmddd.domain;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Base class for aggregate roots.
 * Extends Entity with domain event management and aggregate boundary enforcement.
 * 
 * @param <ID> The type of the aggregate root identifier
 */
public abstract class AggregateRoot<ID> extends Entity<ID> {
    
    private final List<DomainEvent> domainEvents;
    
    protected AggregateRoot(ID id, String domainContext) {
        super(id, domainContext);
        this.domainEvents = new ArrayList<>();
    }
    
    /**
     * Add a domain event to be published
     * @param event The domain event to add
     */
    protected void addDomainEvent(DomainEvent event) {
        if (event != null) {
            domainEvents.add(event);
        }
    }
    
    /**
     * Get pending domain events
     * @return Unmodifiable list of pending domain events
     */
    public List<DomainEvent> getDomainEvents() {
        return Collections.unmodifiableList(domainEvents);
    }
    
    /**
     * Clear domain events after publishing
     */
    public void clearDomainEvents() {
        domainEvents.clear();
    }
    
    /**
     * Get aggregate boundaries definition
     * @return Aggregate boundaries
     */
    public abstract AggregateBoundaries getAggregateBoundaries();
    
    /**
     * Validate aggregate-specific rules
     * @return Validation result for aggregate rules
     */
    protected abstract ValidationResult validateAggregateRules();
    
    @Override
    public ValidationResult validateDomainInvariants() {
        ValidationResult entityValidation = super.validateDomainInvariants();
        ValidationResult aggregateValidation = validateAggregateRules();
        
        ValidationResult combined = new ValidationResult();
        
        // Combine validation results
        if (!entityValidation.isValid()) {
            entityValidation.getErrors().forEach(combined::addError);
        }
        
        if (!aggregateValidation.isValid()) {
            aggregateValidation.getErrors().forEach(combined::addError);
        }
        
        entityValidation.getWarnings().forEach(combined::addWarning);
        aggregateValidation.getWarnings().forEach(combined::addWarning);
        
        return combined;
    }
    
    @Override
    public CompletableFuture<ModuleHealth> getModuleStatus() {
        return super.getModuleStatus().thenApply(health -> {
            // Enhance with aggregate-specific information
            health.getHealthIndicators().put("pending_events", domainEvents.size());
            health.getHealthIndicators().put("aggregate_type", getClass().getSimpleName());
            return health;
        });
    }
    
    /**
     * Apply business operation and generate domain events
     * This is a template method that subclasses can override
     * @param operation The business operation to apply
     * @return CompletableFuture that completes when operation is applied
     */
    protected CompletableFuture<Void> applyBusinessOperation(BusinessOperation operation) {
        return CompletableFuture.runAsync(() -> {
            // Template method - subclasses should override
            updateVersion();
        });
    }
    
    /**
     * Marker interface for business operations
     */
    public interface BusinessOperation {
        String getOperationName();
    }
}

File: stubs/csharp/src/BeastMode.RmDdd/Domain/AggregateRoot.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using BeastMode.RmDdd.Core;
using BeastMode.RmDdd.Utilities;

namespace BeastMode.RmDdd.Domain
{
    /// <summary>
    /// Base class for aggregate roots.
    /// Extends Entity with domain event management and aggregate boundary enforcement.
    /// </summary>
    /// <typeparam name="TId">The type of the aggregate root identifier</typeparam>
    public abstract class AggregateRoot<TId> : Entity<TId>
        where TId : notnull
    {
        private readonly List<DomainEvent> _domainEvents = new();
        
        protected AggregateRoot(TId id, string domainContext) : base(id, domainContext)
        {
        }
        
        /// <summary>
        /// Add a domain event to be published
        /// </summary>
        /// <param name="domainEvent">The domain event to add</param>
        protected void AddDomainEvent(DomainEvent domainEvent)
        {
            if (domainEvent != null)
            {
                _domainEvents.Add(domainEvent);
            }
        }
        
        /// <summary>
        /// Get pending domain events
        /// </summary>
        /// <returns>Read-only list of pending domain events</returns>
        public IReadOnlyList<DomainEvent> GetDomainEvents()
        {
            return _domainEvents.AsReadOnly();
        }
        
        /// <summary>
        /// Clear domain events after publishing
        /// </summary>
        public void ClearDomainEvents()
        {
            _domainEvents.Clear();
        }
        
        /// <summary>
        /// Get aggregate boundaries definition
        /// </summary>
        /// <returns>Aggregate boundaries</returns>
        public abstract AggregateBoundaries GetAggregateBoundaries();
        
        /// <summary>
        /// Validate aggregate-specific rules
        /// </summary>
        /// <returns>Validation result for aggregate rules</returns>
        protected abstract ValidationResult ValidateAggregateRules();
        
        public override ValidationResult ValidateDomainInvariants()
        {
            var entityValidation = base.ValidateDomainInvariants();
            var aggregateValidation = ValidateAggregateRules();
            
            return entityValidation.Combine(aggregateValidation);
        }
        
        public override async Task<ModuleHealth> GetModuleStatusAsync(CancellationToken cancellationToken = default)
        {
            var baseHealth = await base.GetModuleStatusAsync(cancellationToken);
            var healthIndicators = new Dictionary<string, object>(baseHealth.HealthIndicators)
            {
                ["pending_events"] = _domainEvents.Count,
                ["aggregate_type"] = GetType().Name
            };
            
            return new ModuleHealth(
                baseHealth.Status,
                baseHealth.Message,
                baseHealth.Capabilities,
                baseHealth.DomainHealth,
                healthIndicators
            );
        }
        
        /// <summary>
        /// Apply business operation and generate domain events
        /// This is a template method that subclasses can override
        /// </summary>
        /// <param name="operation">The business operation to apply</param>
        /// <param name="cancellationToken">Cancellation token</param>
        /// <returns>Task that completes when operation is applied</returns>
        protected virtual Task ApplyBusinessOperationAsync(IBusinessOperation operation, CancellationToken cancellationToken = default)
        {
            // Template method - subclasses should override
            UpdateVersion();
            return Task.CompletedTask;
        }
    }
    
    /// <summary>
    /// Marker interface for business operations
    /// </summary>
    public interface IBusinessOperation
    {
        /// <summary>
        /// Gets the name of the operation
        /// </summary>
        string OperationName { get; }
    }
}